### PR TITLE
Trying matching and scaling

### PR DESCRIPTION
**Pull request by nicolaskruchten**
Resolves https://github.com/plotly/plotly.js/issues/3539.

@plotly/plotly_js

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
Trying here: https://codepen.io/nicolaskruchten/pen/KKMPJKR ... no luck yet.

**Comment by nicolaskruchten** [comment]
@emmanuelle check out the codepen above as a proof of concept

File: src/plots/cartesian/constraints.js [hunk 1]
```diff
@@ -57,10 +57,7 @@ exports.handleConstraintDefaults = function(containerIn, containerOut, coerce, o
     // 'matches' wins over 'scaleanchor' (for now)
     var scaleanchor, scaleOpts;
 
-    if(!matches &&
```
Comment (nicolaskruchten) [inline src/plots/cartesian/constraints.js:60]: @alexcjohnson so I loosened this check, but I also had to remove the check below ... is this a belt-and-suspenders situation? If I restore the check below such that we only remove scale if it's the same letter as matches, does that seem sufficient?

**Comment by archmoj** [comment]
DEMO: [before](https://codepen.io/MojtabaSamimi/pen/ZEOeMJe?editors=0010) vs [after](https://codepen.io/MojtabaSamimi/pen/YzWZOLB?editors=1000).

**Comment by alexcjohnson** [comment]
Thanks for the demo @archmoj 

AFAICT this generally works OK for `constrain: 'domain'` if you specify every possible constraint you can - which is generally quite overconstrained so it's clearly going to be easy to specify impossible constraints. It's also clearly broken for `constrain: 'range'` and strange things happen if you try to remove some of the redundant constraints.

`fixedrange` currently seems to have no effect on the ranges we determine at render time, but if users zoom or pan afterward the constraints are sometimes violated as the `fixedrange` axis does not change its range, and sometimes this axis doesn't change while the mouse is down but it does after the mouse is released. The constraint violation doesn't bother me except that if you tried to take the resulting figure and recreate the plot in a new container it would want to change *something* to satisfy the constraint.

Having played with this a bit and tried to figure out as a user how I'd construct a plot like this, I really think we should keep the original condition that setting `matches` precludes `scaleanchor` and `scaleratio`, ie there's only one constraint defined per axis, and just allow mixed `matches` and `scaleanchor` constraints. Then to make a plot with many matching aspect-ratio-constrained subplots, you'd set a single `scaleanchor` constraint (eg `yaxis.scaleanchor = 'x'`) and many `matches` constraints (`xaxisN.matches = 'x'`, `yaxisN.matches = 'y'`). For this to mean that the aspect ratios of all of these subplots are identical, the user needs to ensure that the `matches` axes all have the same size domain; But to me that feels like it would be intuitive to users (and mostly hidden by px anyway 😄 )

One important observation from my testing just now: if you have a domain-constrained axis, and others match it, we need to set them all to have the same fractional domain reduction - ie `matches` implies not just the same range, but also inheriting the `constrain` attribute (I suppose `constraintoward` could be different though) and during rendering reducing the range of one of these axes reduces the domain of all of them.

**Comment by alexcjohnson** [comment]
Superseded by #5287

**Pull request closed without merging**